### ISSUE TITLE
[Monitoring V2] multi-node cluster only registers 1 node for kube-etcd, kube-proxy -> many targets down in prometheus for RKE2 cluster using calico

### ISSUE THREAD

**Issue by slickwarren** [rancher/rancher#33052]
**What kind of request is this (question/bug/enhancement/feature request):**
* bug

**Steps to reproduce (least amount of steps as possible):**
* deploy downstream rke2 cluster (3etcd, 1controlplane, 3worker)
* install monitoring on said cluster

**Result:**
* the following targets are completely down in prometheus:
  * rancher-monitoring-kube-scheduler/0 (0/1 up)
  * rancher-monitoring-kube-controller-manager/0 (0/1 up)
* the following targets only have 1 instance up:
  * rancher-monitoring-kube-etcd/0 (1/3 up)
  * rancher-monitoring-kube-proxy/0 (1/7 up)

**Other details that may be helpful:**
error in logs of kube-proxy, kube-controller, kube-etcd and kube-scheduler all look similar, repeating errors like this:
```
level=info ts=2021-06-04T16:16:49.177Z caller=coordinator.go:116 msg=DoScrape scrape_id=23fea160-7da8-4748-a739-2a23474ddfbf url=http://143.198.130.62:10249/metrics
level=error ts=2021-06-04T16:16:59.171Z caller=main.go:177 msg="Error scraping:" err="Timeout reached for \"http://143.198.130.62:10249/metrics\": context canceled" url=http://143.198.130.62:10249/metrics
```

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI):master-head (772ef96)
- Installation option (single install/HA):single

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported):DO
- Machine type (cloud/VM/metal) and specifications (CPU/memory):2vcpu, 7 nodes
- Kubernetes version (use `kubectl version`):v1.21.1-rc2+rke2r1

**Comment by aiyengar2** [comment]
Cannot provision a cluster to replicate this setup due to https://github.com/rancher/rancher/issues/33464.

**Comment by aiyengar2** [comment]
Talked to @slickwarren offline and he was able to reproduce this issue in a 5 node cluster with all roles assigned to all nodes. Therefore, this issue should not be blocked on https://github.com/rancher/rancher/issues/33464 for development, but will be blocked for validating the fix.

**Comment by aiyengar2** [comment]
Was able to reproduce this in a 5 node cluster with all roles assigned to all nodes.

---

### What is the issue?

It seems like the root cause of this issue has something to do with the `rke2-coredns-rke2-coredns-xxxx` pod not setting up DNS correctly on all hosts; I noticed that every time I deleted the pod and got it scheduled onto a new node, the node that the coredns pod was rescheduled to would be "magically" fixed for scrapes. This was consistently reproducible with multiple tries.

This hypothesis is further supported by the fact that the following command produced different results on the nodes where coredns was not deployed v.s. where coredns was deployed:

#### Nodes where coredns was not deployed

```bash
$ wget http://pushprox-kube-controller-manager-proxy.cattle-monitoring-system.svc
wget: bad address 'pushprox-kube-controller-manager-proxy.cattle-monitoring-system.svc'
```

#### Node where coredns was deployed

```bash
$ wget http://pushprox-kube-controller-manager-proxy.cattle-monitoring-system.svc
Connecting to pushprox-kube-controller-manager-proxy.cattle-monitoring-system.svc (10.43.178.111:80)
```

---

### Why is this DNS resolution step necessary?

PushProx works by having the client establish an outbound connection with a proxy. The proxy collects the scrape requests and the client "polls" for any requests that need to be satisfied and accordingly executes them.

In the logs for the failing pushprox clients, we see the following log that indicates that it failed to make an outbound connection with the proxy (since the proxy's DNS is not resolvable)

```
level=error ts=2021-07-09T00:05:24.423Z caller=main.go:222 msg="Error polling:" err="Post \"http://pushprox-kube-controller-manager-proxy.cattle-monitoring-system.svc:8080/poll\": dial tcp: i/o timeout"
```

Similarly, in the logs of the proxy, we routinely see that while 5 `DoScrape` requests are sent to the proxy by Prometheus, only one proxy has gotten back a response (`Responded to /poll`) and executes the scrape request (`WaitForScrapeInstruction` and `Got /push`). The other requests result in `Error scraping` and the GC log shows that only 1 client is registered (`GC of clients completed" deleted=0 remaining=1`):

```
Thu, Jul 8 2021 5:14:56 pm | level=info ts=2021-07-09T00:14:56.021Z caller=coordinator.go:116 msg=DoScrape scrape_id=732fd60a-d485-4043-b4e4-8dff19b54782 url=http://143.198.155.55:10252/metrics
...
Thu, Jul 8 2021 5:13:27 pm | level=info ts=2021-07-09T00:13:27.443Z caller=coordinator.go:116 msg=DoScrape scrape_id=f77b1073-4223-4ec4-b131-52b9b68668e6 url=http://143.198.148.70:10252/metrics
Thu, Jul 8 2021 5:13:32 pm | level=info ts=2021-07-09T00:13:32.028Z caller=coordinator.go:116 msg=DoScrape scrape_id=0580e1fa-06c8-4e3e-b7d2-2df057d3c8ce url=http://143.198.145.66:10252/metrics
Thu, Jul 8 2021 5:13:34 pm | level=info ts=2021-07-09T00:13:34.278Z caller=coordinator.go:116 msg=DoScrape scrape_id=6d8a981e-ffe3-42c1-8e76-784b611724e4 url=http://164.90.159.208:10252/metrics
Thu, Jul 8 2021 5:13:35 pm | level=info ts=2021-07-09T00:13:35.457Z caller=coordinator.go:116 msg=DoScrape scrape_id=ed7b6347-267b-409a-bb13-b6483698b40d url=http://143.198.154.186:10252/metrics
Thu, Jul 8 2021 5:13:35 pm | level=info ts=2021-07-09T00:13:35.458Z caller=main.go:154 msg="Responded to /poll" url=http://143.198.154.186:10252/metrics scrape_id=ed7b6347-267b-409a-bb13-b6483698b40d
Thu, Jul 8 2021 5:13:35 pm | level=info ts=2021-07-09T00:13:35.459Z caller=coordinator.go:137 msg=WaitForScrapeInstruction fqdn=143.198.154.186
Thu, Jul 8 2021 5:13:35 pm | level=info ts=2021-07-09T00:13:35.462Z caller=main.go:136 msg="Got /push" scrape_id=ed7b6347-267b-409a-bb13-b6483698b40d
Thu, Jul 8 2021 5:13:35 pm | level=info ts=2021-07-09T00:13:35.462Z caller=coordinator.go:169 msg=ScrapeResult scrape_id=ed7b6347-267b-409a-bb13-b6483698b40d
Thu, Jul 8 2021 5:13:37 pm | level=error ts=2021-07-09T00:13:37.439Z caller=main.go:177 msg="Error scraping:" err="Timeout reached for \"http://143.198.148.70:10252/metrics\": context canceled" url=http://143.198.148.70:10252/metrics
Thu, Jul 8 2021 5:13:42 pm | level=error ts=2021-07-09T00:13:42.024Z caller=main.go:177 msg="Error scraping:" err="Timeout reached for \"http://143.198.145.66:10252/metrics\": context canceled" url=http://143.198.145.66:10252/metrics
Thu, Jul 8 2021 5:13:44 pm | level=error ts=2021-07-09T00:13:44.273Z caller=main.go:177 msg="Error scraping:" err="Timeout reached for \"http://164.90.159.208:10252/metrics\": context canceled" url=http://164.90.159.208:10252/metrics
Thu, Jul 8 2021 5:13:51 pm | level=info ts=2021-07-09T00:13:51.006Z caller=coordinator.go:221 msg="GC of clients completed" deleted=0 remaining=1
```

**Comment by aiyengar2** [comment]
Note: This issue does **not** impact Monitoring V2 deployed in **imported RKE2** clusters. 

It only affects Monitoring V2 deployed in RKE2 clusters provisioned by Rancher.

---

Just tested this in the latest master-head and I observe that the targets are all up in an **imported** RKE2 cluster with version `v1.21.2+rke2r1`.

![Screen Shot 2021-07-09 at 3 43 53 PM](https://user-images.githubusercontent.com/13127443/125142393-78d35c80-e0cc-11eb-9708-69257b1e13af.png)

![targets_up](https://user-images.githubusercontent.com/13127443/125142508-c5b73300-e0cc-11eb-8e88-0cab78840deb.png)

Targets are also all up in an **imported** RKE2 cluster with version `v1.20.8+rke2r1`:

![Screen Shot 2021-07-09 at 3 47 02 PM](https://user-images.githubusercontent.com/13127443/125142562-e8e1e280-e0cc-11eb-8217-a8c50f4b7fef.png)

![targets-up](https://user-images.githubusercontent.com/13127443/125142738-4d04a680-e0cd-11eb-83a2-7934c287d77e.png)

**Comment by aiyengar2** [comment]
In `3ae6f43` master-head, I no longer see this issue in provisioned clusters either. However, after investigation, I'm unsure what would have caused this to be resolved.

Moving back to test.

---

![Screen Shot 2021-07-09 at 4 36 41 PM](https://user-images.githubusercontent.com/13127443/125144799-de771700-e0d3-11eb-86da-5555c4655f9f.png)

![targets_up](https://user-images.githubusercontent.com/13127443/125144817-f2227d80-e0d3-11eb-971a-d4148363b2c6.png)

**Comment by sowmyav27** [comment]
@kravciak Can you validate this using DO RKE2 provisioning  ?
- 1 etcd, 1 control and 1 worker
- 1 etcd, 1 control and 3 workers

**Comment by kravciak** [comment]
Unable to provision DO RKE2 cluster with this config, blocked for now.

![Screenshot from 2021-07-27 17-11-15](https://user-images.githubusercontent.com/538604/127180143-bbfa3bad-a1a8-425a-b560-b88364a34429.png)

**Comment by slickwarren** [comment]
still blocked by [this issue](https://github.com/rancher/rancher/issues/32362#issuecomment-891294221)

**Comment by slickwarren** [comment]
@aiyengar2 I can reproduce this on 2.6-head (222dced) using the same downstream cluster as originally reported (3etcd, 2 cp, 3 worker in Digital Ocean)

**Comment by aiyengar2** [comment]
Yes, it seems like this issue came back. The symptoms seem to be the same as described in https://github.com/rancher/rancher/issues/33052#issuecomment-876824380, which makes me suspect that this is an RKE2 regression.

- Broken on `v1.21.1-rc2+rke2r1`, as shown in https://github.com/rancher/rancher/issues/33052#issuecomment-876824380
- Fixed on `v1.21.2+rke2r1`, as shown in https://github.com/rancher/rancher/issues/33052#issuecomment-877509776
- Broken on `v1.21.3-rc3+rke2r2` now

**Comment by aiyengar2** [comment]
Seems like this might be related to an RKE2 issue. Opened up a ticket in https://github.com/rancher/rke2/issues/1541 to follow up

**Comment by aiyengar2** [comment]
In Rancher v2.6-head `4b6af95` in the latest RKE2 version (`v1.21.3-rc4+rke2r2`), this issue is **not** observed in RKE2 canal clusters. This issue only affects RKE2 clusters that use calico as the CNI.

### No Unhealthy Targets
![147 182 255 80_k8s_clusters_c-m-926d52a5_api_v1_namespaces_cattle-monitoring-system_services_http_rancher-monitoring-prometheus_9090_proxy_targets](https://user-images.githubusercontent.com/13127443/128751326-92245577-5dca-4e81-8433-0bbcb3fc1b37.png)

### All Targets Are Up
![147 182 255 80_k8s_clusters_c-m-926d52a5_api_v1_namespaces_cattle-monitoring-system_services_http_rancher-monitoring-prometheus_9090_proxy_targets (1)](https://user-images.githubusercontent.com/13127443/128751331-257518c6-22af-4187-8091-b884262b901a.png)


cc: @deniseschannon

**Comment by deniseschannon** [comment]
@slickwarren @sowmyav27 We will release note as a known issue and we are following up with Tigera on a fix.

**Comment by aiyengar2** [comment]
Based on the milestone in https://github.com/rancher/rke2/issues/1541, we should be able to move this back to test in `v1.21.4+rke2r1 `.

@manuelbuil thanks for the quick turnaround! 😄

**Comment by aiyengar2** [comment]
Tested on Rancher master-head `121323a` on a downstream RKE2 cluster with version `v1.21.3-rc5+rke2r2` and I'm unable to reproduce:

##### Cluster/Node Information

![157 230 183 111_dashboard_c_c-m-c86307eb_explorer_node](https://user-images.githubusercontent.com/13127443/129055604-61ee34bc-0bcb-46ea-b92e-0f48cef4b816.png)

##### All Targets Are Healthy

![157 230 183 111_k8s_clusters_c-m-c86307eb_api_v1_namespaces_cattle-monitoring-system_services_http_rancher-monitoring-prometheus_9090_proxy_targets](https://user-images.githubusercontent.com/13127443/129055624-2695c7ad-ea69-4535-a6ec-6f11d2a0fc9e.png)

##### No Unhealthy Targets

![157 230 183 111_k8s_clusters_c-m-c86307eb_api_v1_namespaces_cattle-monitoring-system_services_http_rancher-monitoring-prometheus_9090_proxy_targets (1)](https://user-images.githubusercontent.com/13127443/129055650-8ad414bb-2301-4155-bbee-6671c0a10854.png)

---

cc: @slickwarren ready to be reproduced and closed 😄

**Comment by slickwarren** [comment]
I can confirm, on v2.6-head (9d2e086) and master-head (4414103) multi-node RKE2 clusters now have all targets up after installing monitoring v2, 100.0.0+up16.6.0:
* dedicated nodes (original report: 3etcd, 2cp, 3wk) -- pass
* 3 all-role nodes -- pass

**Comment by slickwarren** [comment]
I’m seeing the issue with monitoringv2 with targets down in a multi cluster setup. using v2.6.0-rc4
The cluster is proxied , and the following only have 1 node up:
`kube-etcd, kube-scheduler , kube-proxy, kube-controller-manager`

I did just bring up a non-proxied cluster and don’t see this issue there, so I think this is now specific to multi-node clusters behind a proxy. Arvind has identified this as the root issue: https://github.com/rancher/rke2/issues/1541

reproduce the issue: 
* bring up a cluster using private nodes, using a proxy (configured both on the nodes and in rancher's registration cmd)
  * I used a 3 node, all roles cluster
* wait for cluster to provision, then install monitoring v2
  * check targets in prometheus -> 1/3 are up for the aforementioned targets

**Comment by rancher-max** [comment]
This should be fixed in v1.21.3-rc7+rke2r2 per https://github.com/rancher/rke2/issues/1541#issuecomment-904112597. @slickwarren I'm going to move this back to test for you to take a look.

**Comment by aiyengar2** [comment]
@minhtt159 this issue was tracking the effort in Rancher to verify if the fix in https://github.com/rancher/rke2/issues/1541 resolved issues in Monitoring. Based on the information you provided, it's possible that RKE2 may have regressed on that issue.

Can you try to follow the steps detailed in the RKE2 issue to reproduce (e.g. run the `dig` command and see if it fails) and, if it is reproduced, comment on the RKE2 issue? There was nothing done from the Rancher side iirc for this issue, so any fix should resolve after upgrading to a new RKE2 version, if you are tracking the same root cause as this ticket.

If you can't reproduce the RKE2 issue, I would recommend opening up a new ticket since it's probably a different root cause.
